File: 2024/13/Go/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"path/filepath"
	"strconv"
	"strings"
)

type Machine struct {
	a, b, p, p2 []int
}

func calc(A [2][2]int, b [2]int) [2]float64 {
	det := A[0][0]*A[1][1] - A[0][1]*A[1][0]
	if det == 0 {
		return [2]float64{0, 0}
	}
	return [2]float64{
		float64(b[0]*A[1][1]-b[1]*A[0][1]) / float64(det),
		float64(A[0][0]*b[1]-A[1][0]*b[0]) / float64(det),
	}
}

func (m *Machine) cost(part2 bool) int {
	var p []int
	if part2 {
		p = m.p2
	} else {
		p = m.p
	}

	result := calc(
		[2][2]int{{m.a[0], m.b[0]}, {m.a[1], m.b[1]}},
		[2]int{p[0], p[1]},
	)

	x, y := result[0], result[1]
	if isInteger(x) && isInteger(y) {
		return int(x)*3 + int(y)
	}
	return 0
}

func isInteger(value float64) bool {
	return value == float64(int(value))
}

func parseNumbers(line string) []int {
	numStrs := strings.FieldsFunc(line, func(r rune) bool {
		return r < '0' || r > '9'
	})
	nums := make([]int, len(numStrs))
	for i, numStr := range numStrs {
		num, _ := strconv.Atoi(numStr)
		nums[i] = num
	}
	return nums
}

func main() {
	file, err := os.Open(filepath.Join("..", "input.txt"))
	if err != nil {
		panic(err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	var input [][]string
	var lines []string
	for scanner.Scan() {
		line := scanner.Text()
		if line == "" {
			input = append(input, lines)
			lines = nil
		} else {
			lines = append(lines, line)
		}
	}
	if len(lines) > 0 {
		input = append(input, lines)
	}

	var machines []Machine
	for _, lines := range input {
		machine := Machine{
			a: parseNumbers(lines[0]),
			b: parseNumbers(lines[1]),
			p: parseNumbers(lines[2]),
		}
		machine.p2 = make([]int, len(machine.p))
		for i, x := range machine.p {
			machine.p2[i] = x + int(1e13)
		}
		machines = append(machines, machine)
	}

	sum := 0
	for _, m := range machines {
		sum += m.cost(false)
	}
	fmt.Println(sum)

	sumPart2 := 0
	for _, m := range machines {
		sumPart2 += m.cost(true)
	}
	fmt.Println(sumPart2)
}

File: 2024/13/Node/main.js
const fs = require('fs');
const path = require('path');

const input = fs
  .readFileSync(path.join(__dirname, '../input.txt'), 'utf-8')
  .split('\n\n')
  .map((line) => line.split('\n'));

function calc(A, b) {
  const det = A[0][0] * A[1][1] - A[0][1] * A[1][0];
  if (det === 0) return [0, 0];
  return [
    (b[0] * A[1][1] - b[1] * A[0][1]) / det,
    (A[0][0] * b[1] - A[1][0] * b[0]) / det,
  ];
}

class Machine {
  constructor(lines) {
    this.a = (lines[0].match(/\d+/g) || []).map(Number);
    this.b = (lines[1].match(/\d+/g) || []).map(Number);
    this.p = (lines[2].match(/\d+/g) || []).map(Number);
    this.p2 = this.p.map((x) => x + 1e13);
  }

  cost(part2 = false) {
    const [x, y] = calc(
      [
        [this.a[0], this.b[0]],
        [this.a[1], this.b[1]],
      ],
      part2 ? this.p2 : this.p,
    );
    return Number.isInteger(x) && Number.isInteger(y) ? x * 3 + y : 0;
  }
}

const machines = input.map((lines) => new Machine(lines));
console.log(machines.reduce((sum, m) => sum + m.cost(), 0));
console.log(machines.reduce((sum, m) => sum + m.cost(true), 0));
